### PR TITLE
[3.2] Implement @at-root

### PR DESCRIPTION
**Pull request by xzyfer**
This PR implements `@at-root`.

### TODO

- [x] [basic](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/at-root/basic)
- [x] [nested](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/at-root/nested)
- [x] [@media](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/at-root/media)
- [x] [@keyframes](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/at-root/keyframes)
- [x] [@entends](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/at-root/extend)
- [x] [&](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/at-root/ampersand)
- [x] [with and without](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/at-root/with_without)
- [x] maintain the correct indentation for nested output mode

This feature is good to go. 

Fixes https://github.com/sass/libsass/issues/353. Specs added https://github.com/sass/sass-spec/pull/175, https://github.com/sass/sass-spec/pull/217, https://github.com/sass/sass-spec/pull/219.

***
It turns out there were a lot of todo tests lying around addressing at-root. As a result the following specs now pass with this PR.

- [135_test_simple_at_root](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/135_test_simple_at_root)
- [136_test_at_root_with_selector](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/136_test_at_root_with_selector)
- [137_test_at_root_in_mixin](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/137_test_at_root_in_mixin)
- [138_test_at_root_in_media](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/138_test_at_root_in_media)
- [139_test_at_root_in_bubbled_media](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/139_test_at_root_in_bubbled_media)
- [140_test_at_root_in_unknown_directive](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/140_test_at_root_in_unknown_directive) 
- [141_test_at_root_with_parent_ref](https://github.com/sass/sass-spec/tree/master/spec/ibsass-todo-tests/scss-tests/141_test_at_root_with_parent_ref)
- [142_test_multi_level_at_root_with_parent_ref](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/scss-tests/142_test_multi_level_at_root_with_parent_ref)
- [143_test_multi_level_at_root_with_inner_parent_ref](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/scss-tests/143_test_multi_level_at_root_with_inner_parent_ref)

### CODE REVIEW

**Comment by xzyfer** [comment]
This is currently passing all the at-root specs. I'll do some clean but otherwise it's ready to go with #790 is merged.

File: prelexer.cpp [hunk 1]
```diff
@@ -213,6 +213,22 @@ namespace Sass {
...
+
+    const char* with_directive(const char* src) {
+      return exactly<with_kwd>(src);
+    }
+
+    const char* without_directive(const char* src) {
+      return exactly<without_kwd>(src);
+    }
+
+    const char* foo(const char* src) {
```
Comment (xzyfer) [inline prelexer.cpp]: Rename this.

File: to_string.hpp [hunk 2]
```diff
@@ -1,3 +1,4 @@
+#ifndef SASS_TO_STRING
```
Comment (xzyfer) [inline to_string.hpp]: This can probably go in it's own PR.

File: parser.cpp [hunk 3]
```diff
@@ -1746,6 +1756,60 @@ namespace Sass {
...
+      body = new (ctx.mem) Block(r->path(), r->position(), 1);
+      *body << r;
+    }
+    in_at_root = false;
+    At_Root_Block* at_root = new (ctx.mem) At_Root_Block(path, at_source_position, body);
+    if (expr) at_root->expression(expr);
+    return at_root;
+  }
+
+  At_Root_Expression* Parser::parse_at_root_expression()
```
Comment (xzyfer) [inline parser.cpp]: I'm want to experiment with how this looks with the `At_Root_Expression`. I don't think that AST node makes a lot of sense.

File: cssize.cpp [hunk 4]
```diff
@@ -105,6 +105,101 @@ namespace Sass {
...
+    At_Root_Block* mm = new (ctx.mem) At_Root_Block(m->path(),
+                                                    m->position(),
+                                                    wrapper_block,
+                                                    m->expression());
+
+    Bubble* bubble = new (ctx.mem) Bubble(mm->path(), mm->position(), mm);
+    return bubble;
+  }
+
+  inline Statement* Cssize::bubble(Feature_Block* m)
```
Comment (xzyfer) [inline cssize.cpp]: This behaviour can probably come out into it's own PR.

File: parser.cpp (hunk 3 above)
Comment (xzyfer) [inline parser.cpp]: I'm opting out of this since managing the specs will be complicated.

File: cssize.cpp (hunk 4 above)
Comment (xzyfer) [inline cssize.cpp]: I'm opting out of this since managing the specs will be complicated.

**Comment by xzyfer** [comment]
I managed to rebase this post-sourcemaps-mega-merge with relatively little pain. I now have a failing at-root spec though. Once that's resolved I'll ship this.

**Comment by mgreter** [comment]
:+1:

**Comment by xzyfer** [comment]
This is ready to go pending CI. I corrected some issues with nested output.

**Comment by xzyfer** [comment]
Thanks for catching the output stuff @mgreter I amended that commit. That has also broken generic at-rule output (more than I expected).  I was going to address at-rule bubbling and output in a different PR but I'll have to handle it here.
